c/prima.c

```diff
@@ -80,59 +80,13 @@ int prima_check_problem(prima_problem_t *problem, prima_options_t *options, cons
     return PRIMA_NULL_X0;
   if ((use_constr && !problem->calcfc) || (!use_constr && !problem->calfun))
     return PRIMA_NULL_FUNCTION;
-  if (alloc_bounds && !problem->xl)
-  {
-    problem->xl = (double*)malloc(problem->n * sizeof(double));
-    if (!problem->xl)
-      return PRIMA_MEMORY_ALLOCATION_FAILS;
-    problem->_allocated_xl = 1;
-    for (int i = 0; i< problem->n; ++ i)
-      problem->xl[i] = -INFINITY;
-  }
-  if (alloc_bounds && !problem->xu)
-  {
-    problem->xu = (double*)malloc(problem->n * sizeof(double));
-    if (!problem->xu)
-      return PRIMA_MEMORY_ALLOCATION_FAILS;
-    problem->_allocated_xu = 1;
-    for (int i = 0; i< problem->n; ++ i)
-      problem->xu[i] = INFINITY;
-  }
-  if (options->maxfun < 0)
+  if (options->maxfun == 0)
```

Comment:
Indeed, it is OK to send `maxfun = 0` and `npt = 0` to the Fortran backend, which will interpret them correctly. Recall https://github.com/libprima/prima/pull/135#issuecomment-1897853547 

What do you think?

Reply:
I think that the C code should initialize as much as possible, or it should pass things to Fortran that get interpreted as "not present". There are two reasons why I think this:

1. Debug complexity.

Let's say, as a crude example, someone using PRIMA from C thinks that `maxfun` not being respected, and so they set it to 0. They'll see it's not being respected. If they go look at the C code, they would see it is set to 0 be default, and maybe they don't see the comments because future PRs have moved them or added new features with different comments or the comments are in the header file and they're looking at the source file. In order to find why `maxfun` isn't being respected they have to go through `algorithm_c`, `algorithm`, and finally `preproc`, which is towards the end of `algorithm`. That's a lot of code to go through. If the initialization is in C, they can end their search much more quickly.

Of course this is a crude example but the point is to illustrate how it can be difficult and challenging to debug code if the default value is interpreted many levels away from the place where the default value is set.

2. Increased complexity of the code from the perspective of new users.

Suppose a new user wants to change to how things are initialized. I don't have a crude example for why someone might want to do this but please play along. If they're a Fortran user then when they make modifications to preproc or anywhere else that looks for 0 or NaN as a default value, it won't occur to them that they might have to make changes to the C code as well. When they're told in the PR that they have to consider C as well, they can't simply copy their changes into C and adapt for the language. They need to understand the conventions and then ensure that C is following those conventions. This seems like it's more complicated than just copying initialization code and adapting it to the "local language." The point here is that it can make the code more difficult to understand.

Documentation may help with these both of these, but I see documentation as a backup solution. I've made many times where I've changed code that *I wrote* and yet forgot to change the comment (which I also wrote) right above the line of code. The point is that it's easy to miss documentation, or for it to get out of date. If the solutions can be made in code that tends to be preferable.

With the initialization in C, or with the `algorithm_c` function interpreting the value provided by C as "not present" it's both easier to debug and also easier to understand. For a standalone C implementation the initialization would have to be written in C so I think it's logical to start implementing it in C.

What do you think?

Reply:
> I think that the C code should initialize as much as possible, or it should pass things to Fortran that get interpreted as "not present". There are two reasons why I think this:
> 
> 1. Debug complexity.
> 
> Let's say, as a crude example, someone using PRIMA from C thinks that `maxfun` not being respected, and so they set it to 0. They'll see it's not being respected. If they go look at the C code, they would see it is set to 0 be default, and maybe they don't see the comments because future PRs have moved them or added new features with different comments or the comments are in the header file and they're looking at the source file. In order to find why `maxfun` isn't being respected they have to go through `algorithm_c`, `algorithm`, and finally `preproc`, which is towards the end of `algorithm`. That's a lot of code to go through. If the initialization is in C, they can end their search much more quickly.
> 
> Of course this is a crude example but the point is to illustrate how it can be difficult and challenging to debug code if the default value is interpreted many levels away from the place where the default value is set.
> 
> 2. Increased complexity of the code from the perspective of new users.
> 
> Suppose a new user wants to change to how things are initialized. I don't have a crude example for why someone might want to do this but please play along. If they're a Fortran user then when they make modifications to preproc or anywhere else that looks for 0 or NaN as a default value, it won't occur to them that they might have to make changes to the C code as well. When they're told in the PR that they have to consider C as well, they can't simply copy their changes into C and adapt for the language. They need to understand the conventions and then ensure that C is following those conventions. This seems like it's more complicated than just copying initialization code and adapting it to the "local language." The point here is that it can make the code more difficult to understand.
> 
> Documentation may help with these both of these, but I see documentation as a backup solution. I've made many times where I've changed code that _I wrote_ and yet forgot to change the comment (which I also wrote) right above the line of code. The point is that it's easy to miss documentation, or for it to get out of date. If the solutions can be made in code that tends to be preferable.
> 
> With the initialization in C, or with the `algorithm_c` function interpreting the value provided by C as "not present" it's both easier to debug and also easier to understand. For a standalone C implementation the initialization would have to be written in C so I think it's logical to start implementing it in C.
> 
> What do you think?

I agree to initialize `maxfun` and `npt` as in the current version. Thank you.